### PR TITLE
interpret predictions enhancements

### LINKED ISSUE #735: plot_predictions with random effects
@GStechschulte, we have further played around with plot_predictions and came across some behavior we don't understand . It might be an issue with handling random effects, hence I describe it here: 

For a model with random effects (p(correct, count) ~ 0 + factor3:factor2 + factor1  + (0 + factor3:factor2 | individual)), the predictions seem to be off compared to the data points (e.g. see Factor1=A, Factor2=orange, Factor3=1 but also others):

w_rf

Also, in factor C, the hdi bars get a bit smaller. It's barely visible here, but in another data set (which I cannot share), they are about 3 times smaller than those of levels A and B, apparently without any reason related to the data.

The same model but without the random effects produces predictions which are pretty close to the empirical means:

wo_rf

Many thanks in advance!

Code to reproduce this and the dataset:

```
import pandas as pd
import bambi as bmb
import seaborn as sns

data = pd.read_csv('simulated_data_order_prob2.csv')

priors = {
    "factor3:factor2": bmb.Prior("Normal", mu=0, sigma=1),
    "factor1": bmb.Prior("Normal", mu=0, sigma=1),
    "factor3:factor2|individual": bmb.Prior("Normal", mu=0, sigma=bmb.Prior("Gamma", alpha=3, beta=3))
}

model = bmb.Model(
    "p(correct, count) ~ 0 + factor3:factor2 + factor1  + (0 + factor3:factor2 | individual)",
    data,
    family="binomial",
    categorical=["factor3", "individual", "factor1", "factor2"],
    priors=priors,
    noncentered=False)

idata = model.fit(tune=2000, draws=2000, random_seed=123, init='adapt_diag',
                  target_accept=0.9, idata_kwargs={'log_likelihood':True})

data['report frequency (%)'] = data['correct'] / data['count']

g = sns.catplot(data=data, kind='strip', x='factor3', y='report frequency (%)', 
                hue='factor2', col='factor1', jitter=False, dodge=True)


axs = bmb.interpret.plot_predictions(
    model=model, 
    idata=idata, 
    covariates=["factor3", "factor2", "factor1"],
    pps=False,
    legend=True,
    fig_kwargs={"figsize": (20, 8), "sharey": True},
    prob = .95,
    ax=g.axes
)
```

Data Set: 

[simulated_data_order_prob2.csv](https://github.com/bambinos/bambi/files/12875843/simulated_data_order_prob2.csv)

### PR DESCRIPTION
**Pull request by GStechschulte**
This PR adds new functionality to the `predictions` and `plot_predictions` functions in `interpret` and resolves #735.  Users can now

- compute unit-level predictions
- compute predictions on user-provided values

Previously, users could only pass a string or list of covariates to compute conditional adjusted predictions. Now, `predictions` has "most of" the functionality that {marginaleffects} has. Additionally, the changes result in a more standard API when calling comparisons, predictions, and slopes. Each function has the arg. `conditional` in which the user can "condition" their estimates on. Furthermore, each function can now compute:

- unit-level (user passes `None`)
- user-provided (user passes a dictionary)
- grid of values (user passes a string or list of covariates)

Below, you will find a couple demos:

```python3
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import warnings

import bambi as bmb

data = bmb.load_data('mtcars')
data["cyl"] = data["cyl"].replace({4: "low", 6: "medium", 8: "high"})
data["gear"] = data["gear"].replace({3: "A", 4: "B", 5: "C"})
data["cyl"] = pd.Categorical(data["cyl"], categories=["low", "medium", "high"], ordered=True)

model = bmb.Model("mpg ~ 0 + hp * wt + cyl + gear", data)
idata = model.fit(draws=1000, target_accept=0.95, random_seed=1234)

# unit-level predictions
bmb.interpret.predictions(
    model,
    idata,
    conditional=None
)
```
cyl | gear | hp | wt | estimate | lower_3.0% | upper_97.0%
-- | -- | -- | -- | -- | -- | --
medium | B | 110 | 2.620 | 22.233424 | 20.051966 | 24.476544
medium | B | 110 | 2.875 | 21.320402 | 19.196886 | 23.344765
low | B | 93 | 2.320 | 25.901435 | 24.255648 | 27.558330
medium | A | 110 | 3.215 | 18.751708 | 16.259737 | 21.293185
high | A | 175 | 3.440 | 16.908354 | 15.261489 | 18.666662

Unit level predictions and average over `hp` and `wt` to obtain marginal effects of `gear` and `cyl`:

```python3
bmb.interpret.plot_predictions(
    model,
    idata,
    conditional=None,
    average_by=["gear", "cyl"],
    fig_kwargs={"figsize": (7, 3)},
);
```
image

Compute a pairwise grid using user-provided values and compute predictions:

```python3
bmb.interpret.plot_predictions(
    model,
    idata,
    conditional={
        "hp": [100, 120],
        "cyl": ["low", "medium", "high"],
        "gear": "A",
    },
    subplot_kwargs={"main": "hp", "group": "gear", "panel": "cyl"},
    fig_kwargs={"figsize": (10, 4), "sharey": True},
    legend=True
);
```
image

To do:
- [x] finish updating docs that use `plot_predictions`
- [x] review the added test cases in `plot_predictions`

### CODE REVIEW

**Comment by GStechschulte** [comment]
@tomicapretto although it is ~660 lines of code added 😵‍💫 a lot of it comes from docstring additions and more error handling. Most of the added functionality was leveraged existing functions.

File: docs/notebooks/plot_predictions.ipynb [hunk 1]
```diff
@@ -1,986 +1,1458 @@
 {
```
Comment (tomicapretto) [inline docs/notebooks/plot_predictions.ipynb:1]: <p>What happened here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/bambinos/bambi/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

**Comment by tomicapretto** [comment]
@GStechschulte being 100% honest I can't understand all the changes as well as you do. I do see in many cases you made things more general and you are reusing code more, which is great. So once you finish the implementation, I trust your judgment to merge this.

Two more comments:

1. The test failure is because some non-compatible type hint on Python 3.9. I _think it's fixed [here](https://github.com/pymc-devs/pymc/pull/6945) but I'm not sure.
2. Do you think we can cut a 0.13.0 release after this? I just realized we don't have a release with the `interpret` submodule and I think it's quite mature at this point.

File: docs/notebooks/plot_predictions.ipynb (hunk 1 above)
Comment (GStechschulte) [reply docs/notebooks/plot_predictions.ipynb:1]: If I remember correctly, I interrupted the cell to run another cell in the same notebook. Nothing to worry about here :)

**Comment by tomicapretto** [comment]
Looks great! Go ahead and merge if you don't plan to add anything.

**Comment by jt-lab** [comment]
@GStechschulte, many thanks for the work on the submodule, it really helps a lot! I can report that the preliminary average_by solutions already worked very well not only on the simulated data set but also on others. The predictions align well with the observed data.

**Comment by GStechschulte** [comment]
@jt-lab Thank you and this is great to hear! 😄 We / I really appreciate you taking the time to open the issues and to give feedback. Cheers!

**Pull request merged**